path_login.go

```diff
@@ -0,0 +1,229 @@
+		Pattern: "login/" + framework.GenericNameRegex("role"),
+		Fields: map[string]*framework.FieldSchema{
+			"requestHeaders": {
+				Type:        framework.TypeHeader,
+				Description: `The signed headers of the client`,
+			},
+			"role": {
+				Type:        framework.TypeLowerCaseString,
+				Description: "Name of the role.",
+			},
+		},
+		Callbacks: map[logical.Operation]framework.OperationFunc{
+			logical.UpdateOperation: b.pathLoginUpdate,
+		},
+
+		HelpSynopsis:    pathLoginSyn,
+		HelpDescription: pathLoginDesc,
+	}
+}
+
+func (b *backend) pathLoginUpdate(ctx context.Context, req *logical.Request, data *framework.FieldData) (*logical.Response, error) {
+
+	//Validate the role
+	role, ok := data.GetOk("role")
+	if !ok {
+		return logical.ErrorResponse("Role is not specified"), nil
+	}
+	roleName := role.(string)
+
+	b.Logger().Debug(req.ID, "pathLoginUpdate roleName", roleName)
+
+	//Parse the authentication headers
+	authenticateRequestHeaders := data.Get("requestHeaders").(http.Header)
+
+	//Find the targetUrl and Method
+	finalLoginPath := PathVersionBase + fmt.Sprintf(PathBaseFormat, roleName)
+	method, targetUrl, err := requestTargetToMethodURL(authenticateRequestHeaders[HdrRequestTarget], PathLoginMethod, finalLoginPath)
+	if err != nil {
+		return unauthorizedLogicalResponse(req, b.Logger(), err)
+	}
+	b.Logger().Debug(req.ID, "Method:", method, "targetUrl:", targetUrl)
+
+	authenticateClientDetails := AuthenticateClientDetails{
+		RequestHeaders: authenticateRequestHeaders,
+	}
+
+	requestMetadata := common.RequestMetadata{
+		nil,
+	}
+
+	authenticateClientRequest := AuthenticateClientRequest{
+		authenticateClientDetails,
+		nil,
+		&req.ID,
+		requestMetadata,
+	}
+
+	//Authenticate the request with Identity
+	if b.authenticationClient == nil && b.createAuthClient() != nil {
+		return logical.RespondWithStatusCode(nil, req, http.StatusInternalServerError)
+	}
+	authenticateClientResponse, err := b.authenticationClient.AuthenticateClient(ctx, authenticateClientRequest)
+	if err != nil {
+		return unauthorizedLogicalResponse(req, b.Logger(), err)
+	}
+	if authenticateClientResponse.Principal == nil ||
+		len(authenticateClientResponse.Principal.Claims) == 0 ||
+		*authenticateClientResponse.IsSuccess == false {
+		return unauthorizedLogicalResponse(req, b.Logger(), err)
+	}
+	internalClaims := FromClaims(authenticateClientResponse.Principal.Claims)
+	principalType := internalClaims.GetString(ClaimPrincipalType)
+
+	//Check the principal type
+	if principalType != PrincipalTypeInstance && principalType != PrincipalTypeUser {
+		return unauthorizedLogicalResponse(req, b.Logger(), err)
+	}
+
+	b.Logger().Debug("Authentication ok", "Method:", method, "targetUrl:", targetUrl, "id", req.ID)
+
+	//Validate the home tenancy
+	err = b.validateHomeTenancy(ctx, req, *authenticateClientResponse.Principal.TenantId)
+	if err != nil {
+		return unauthorizedLogicalResponse(req, b.Logger(), err)
+	}
+
+	//Validate that the role exists
```

Comment:
This validation should be moved to the start of the function.

Reply:
Done